File: DSAPython/Graphs/dfs.py
class Graph:
    def __init__(self,nVertices):
        self.nVertices = nVertices
        self.mat = [[] for _ in range(nVertices)]

    def addEdge(self,u,v):
        self.mat[u].append(v)
        self.mat[v].append(u)

    def removeEdge(self,u,v):
        if u in self.mat[v]:
            self.mat[u].remove(v)
            self.mat[v].remove(u)
        else : return

    def __dfsHelper(self,node,vis):
        print(node,end=" ")
        vis[node] = 1

        for nbr in self.mat[node]:
            if vis[nbr] == 0 : self.__dfsHelper(nbr,vis)

    def dfs(self):
        vis = [0] * (self.nVertices)
        for i in range(self.nVertices):
            if vis[i] == 0 : self.__dfsHelper(i,vis)

    def __str__(self):
        return str(self.mat)
    

if __name__ == '__main__':
    g = Graph(7)
    g.addEdge(0,1)
    g.addEdge(1,3)
    g.addEdge(2,4)
    print(g)
    g.removeEdge(1,2)
    print(g)
    g.removeEdge(1,3)
    print(g)
    g.addEdge(2,4)
    g.addEdge(5,6)
    g.dfs()

File: Graphs/DFS.cpp
#include<bits/stdc++.h>
using namespace std;

template <typename T>
class Graph{
    public:
    map<T,list<T>> l;

    void addEdge(int x,int y){
        l[x].push_back(y);
        l[y].push_back(x);
    }

    void dfs_helper(T src,map<T,bool> &visited){
        cout<<src<<" ";
        visited[src]=true;

        for(auto nbr:l[src]){
            if(!visited[nbr]){
                dfs_helper(nbr,visited);
            }
        }
        return;
    }

    void dfs(T src){
        map<T,bool> visited;
        for(auto node:l){
            visited[node.first]=false;
        }
        dfs_helper(src,visited);
    }
};

int main(){
    Graph<int> g;
    g.addEdge(0,1);
    g.addEdge(0,3);
    g.addEdge(1,2);
    g.addEdge(2,3);
    g.addEdge(3,4);
    g.addEdge(4,5);
    g.dfs(0);
    return 0;
}